browser/ui/webui/settings/brave_sync_handler.cc

```diff
@@ -0,0 +1,217 @@
+}
+
+void BraveSyncHandler::OnJavascriptAllowed() {
+  syncer::DeviceInfoTracker* tracker =
+      DeviceInfoSyncServiceFactory::GetForProfile(profile_)
+          ->GetDeviceInfoTracker();
+  DCHECK(tracker);
+  if (tracker)
+    device_info_tracker_observer_.Add(tracker);
+}
+
+void BraveSyncHandler::OnJavascriptDisallowed() {
+  device_info_tracker_observer_.RemoveAll();
+}
+
+void BraveSyncHandler::OnDeviceInfoChange() {
+  if (IsJavascriptAllowed())
+    FireWebUIListener("device-info-changed", GetSyncDeviceList());
+}
+
+void BraveSyncHandler::HandleGetDeviceList(const base::ListValue* args) {
+  AllowJavascript();
+  const auto& list = args->GetList();
+  CHECK_EQ(1U, list.size());
+  const base::Value* callback_id;
+  CHECK(args->Get(0, &callback_id));
+
+  ResolveJavascriptCallback(*callback_id, GetSyncDeviceList());
+}
+
+void BraveSyncHandler::HandleGetSyncCode(const base::ListValue* args) {
+  AllowJavascript();
+
+  CHECK_EQ(1U, args->GetSize());
+  const base::Value* callback_id;
+  CHECK(args->Get(0, &callback_id));
+
+  brave_sync::Prefs brave_sync_prefs(profile_->GetPrefs());
+  std::string sync_code = brave_sync_prefs.GetSeed();
+  if (sync_code.empty()) {
+    std::vector<uint8_t> seed = brave_sync::crypto::GetSeed();
+    sync_code = brave_sync::crypto::PassphraseFromBytes32(seed);
+  }
+
+  ResolveJavascriptCallback(*callback_id, base::Value(sync_code));
+}
+
+void BraveSyncHandler::HandleGetQRCode(const base::ListValue* args) {
+  AllowJavascript();
+  CHECK_EQ(2U, args->GetSize());
+  const base::Value* callback_id;
+  CHECK(args->Get(0, &callback_id));
+  const base::Value* sync_code;
+  CHECK(args->Get(1, &sync_code));
+
+  std::vector<uint8_t> seed;
+  if (!brave_sync::crypto::PassphraseToBytes32(sync_code->GetString(), &seed)) {
+    LOG(ERROR) << "invalid sync code:" << sync_code->GetString();
+    ResolveJavascriptCallback(*callback_id, base::Value(false));
+    return;
+  }
+
+  std::string seed_serialized = StrFromUint8Array(seed);
+
+  // TODO(petemill): Use QRCodeGenerator
+  // const char* input_data_string = sync_code.c_str();
+  // auto input_data = base::span<const uint8_t>(
+  //         reinterpret_cast<const uint8_t*>(input_data_string),
+  //             strlen(input_data_string));
+  // // Generate QR code data
+  // uint8_t qr_data_buf[QRCode::kInputBytes];
+  // QRCodeGenerator generator;
+  // base::span<const uint8_t, QRCodeGenerator::kTotalSize> code =
+  //     generator.Generate(input_data);
+  ResolveJavascriptCallback(*callback_id, base::Value(seed_serialized));
+  // ResolveJavascriptCallback(*callback_id, base::Value(false));
+}
+
+void BraveSyncHandler::HandleSetSyncCode(const base::ListValue* args) {
+  AllowJavascript();
+  CHECK_EQ(2U, args->GetSize());
+  const base::Value* callback_id;
+  CHECK(args->Get(0, &callback_id));
+  const base::Value* sync_code;
+  CHECK(args->Get(1, &sync_code));
+
+  if (sync_code->GetString().empty()) {
+    LOG(ERROR) << "No sync code parameter provided!";
```

Comment:
seems like all of these should be VLOG

Reply:
This one I think should stay as error since our code doesn't supply the wrong number of parameters so if it errors it will be a one off dev error when making a change